Add the --rolebinding-name option to rolebinding commands

mrogers950 opened the pull request:
Add a `--rolebinding-name` option to the rolebinding and clusterrolebinding commands for specifying the name of the rolebinding to modify.

Fixes #13035

stevekuznetsov commented on test/cmd/admin.sh:
```diff
@@ -155,6 +155,30 @@ os::cmd::expect_success 'oadm policy add-cluster-role-to-group cluster-admin sys
 os::cmd::expect_success 'oadm policy remove-cluster-role-from-group cluster-admin system:unauthenticated'
 os::cmd::expect_success 'oadm policy add-cluster-role-to-user cluster-admin system:no-user'
 os::cmd::expect_success 'oadm policy remove-cluster-role-from-user cluster-admin system:no-user'
+# test the rolebinding and clusterrolebinding name flag
```
Please break this out into it's own file under `test/cmd`, We don't need this one to grow any more.

stevekuznetsov commented on test/cmd/admin.sh:
```diff
@@ -155,6 +155,30 @@ os::cmd::expect_success 'oadm policy add-cluster-role-to-group cluster-admin sys
+os::cmd::expect_success 'oadm policy add-role-to-user admin -z fake-sa1 -N rb1'
+os::cmd::expect_success 'oadm policy add-role-to-user admin -z fake-sa2 -N rb2'
+os::cmd::expect_success 'oadm policy add-role-to-user admin -z fake-sa3 -N rb2'
+os::cmd::expect_success_and_text 'oc get rolebinding/rb1' 'rb1       /admin                        fake-sa1'
```
When doing action-state checks with `os::cmd` always prefer to use `jsonpath` to explicitly query the object -- so here, I would expect:

```
os::cmd::expect_success_and_text "oc get rolebinding rb1 -o jsonpath='{ .roleref.name }'" "admin"
```

and the same for the subject

mrogers950 replied:
Is it possible for the jsonpath output to display those three fields from one command? For example, making the output display as 'rb1 /admin fake-sa1'.

stevekuznetsov replied:
It is, when the paths are comma-delimited. See [the doc](https://kubernetes.io/docs/user-guide/jsonpath/)

mrogers950 replied:
Thanks, though I will turn these into unit tests as you suggested.

stevekuznetsov commented on test/cmd/admin.sh:
```diff
@@ -155,6 +155,30 @@ os::cmd::expect_success 'oadm policy add-cluster-role-to-group cluster-admin sys
+os::cmd::expect_success 'oadm policy add-role-to-user admin -z fake-sa3 -N rb2'
+os::cmd::expect_success_and_text 'oc get rolebinding/rb1' 'rb1       /admin                        fake-sa1'
+os::cmd::expect_success_and_text 'oc get rolebinding/rb2' 'rb2       /admin                        fake-sa2, fake-sa3'
+os::cmd::expect_success 'oadm policy remove-role-from-user admin -z fake-sa3 -N rb2'
```
That being said, these look more like unit-level tests rather than cmd integration.

stevekuznetsov replied:
Specifically -- a lot of the create and check flows seem superfluous -- we should be able to check that the struct of `Options` would lead to the correct objects being passed to the client in a unit test. Writing the test like this outside only also checks that the client correctly updates the server, which we don't need to be doing here.

enj commented:
@openshift/cli-review

juanvallejo commented on pkg/cmd/admin/policy/modify_roles.go:
```diff
@@ -371,9 +381,16 @@ subjectCheck:
 	}
 
 	if isUpdate {
+		if o.RoleBindingName != "" {
```
nit: `len(o.RoleBindingName) > 0` when checking for a non-empty empty string

juanvallejo commented on pkg/cmd/admin/policy/policy.go:
```diff
@@ -115,7 +115,7 @@ func getUniqueName(basename string, existingNames *sets.String) string {
 // RoleBindingAccessor is used by role modification commands to access and modify roles
 type RoleBindingAccessor interface {
-	GetExistingRoleBindingsForRole(roleNamespace, role string) ([]*authorizationapi.RoleBinding, error)
+	GetExistingRoleBindingsForRole(roleNamespace, role, rb string) ([]*authorizationapi.RoleBinding, error)
```
I think this might benefit from introducing a new method instead

mrogers950 replied:
The new method would replace both uses of GetExistingRoleBindingsForRole(), so maybe it should be renamed. Eg., GetRoleBindingByRoleNames().

stevekuznetsov replied:
Not sure this parameter makes sense for this method -- also, `s/rb/roleBindingName/`

juanvallejo commented on pkg/cmd/admin/policy/policy.go:
```diff
@@ -186,21 +178,28 @@ func NewClusterRoleBindingAccessor(client client.Interface) ClusterRoleBindingAc
+	ret := make([]*authorizationapi.RoleBinding, 0)
+
+	for _, currBinding := range bindings {
+		if currBinding.RoleRef.Name == role && (rb == "" || rb == currBinding.Name) {
```
nit: `len(rb) == 0`

juanvallejo commented on pkg/cmd/admin/policy/modify_roles.go:
```diff
@@ -71,6 +72,7 @@ func NewCmdAddRoleToGroup(name, fullName string, f *clientcmd.Factory, out io.Wr
 		},
 	}
 
+	cmd.Flags().StringVarP(&options.RoleBindingName, "rolebinding-name", "N", "", "name of the rolebinding to modify or create")
```
`name of the rolebinding to modify or create. If left empty, overrides the first rolebinding found for the given role`

enj replied:
`s/overrides/appends to/`

enj replied:
@openshift/cli-review is it normal to have capital parameters like `-N`?

juanvallejo replied:
We have a few cases of this, such as in [new-app](https://github.com/openshift/origin/blob/master/pkg/cmd/cli/cmd/newapp.go#L229), [rsh](https://github.com/openshift/origin/blob/master/pkg/cmd/cli/cmd/rsh.go#L99), and [new-build](https://github.com/openshift/origin/blob/master/pkg/cmd/cli/cmd/newbuild.go#L126)

liggitt replied:
avoid claiming new single-letter flags, per https://github.com/kubernetes/community/blob/master/contributors/devel/kubectl-conventions.md#flag-conventions

enj commented on pkg/cmd/admin/policy/modify_roles.go:
```diff
@@ -388,7 +405,7 @@ subjectCheck:
 
 func (o *RoleModificationOptions) RemoveRole() error {
-	roleBindings, err := o.RoleBindingAccessor.GetExistingRoleBindingsForRole(o.RoleNamespace, o.RoleName)
+	roleBindings, err := o.RoleBindingAccessor.GetExistingRoleBindingsForRole(o.RoleNamespace, o.RoleName, o.RoleBindingName)
```
Specifying `RoleBindingName` here seems strange to me.  If you are removing a role, you probably want to remove it from all bindings.  Otherwise you risk the user continuing to have the permission you just tried to remove.

Miciah commented on pkg/cmd/admin/policy/modify_roles.go:
```diff
@@ -231,6 +238,7 @@ func NewCmdRemoveClusterRoleFromGroup(name, fullName string, f *clientcmd.Factor
 		},
 	}
 
+	cmd.Flags().StringVarP(&options.RoleBindingName, "rolebinding-name", "N", "", "name of the rolebinding to modify or create")
```
`remove-cluster-role-from-group` doesn't create.

mrogers950 commented:
* Replaced cmd tests with unit tests
* Use default role name behavior for deletes
* Use only --rolebinding-name not -N
* Revised option description, other nits.

stevekuznetsov commented on pkg/cmd/admin/policy/policy.go:
```diff
@@ -131,21 +131,13 @@ func NewLocalRoleBindingAccessor(bindingNamespace string, client client.Interfac
 }
 
-func (a LocalRoleBindingAccessor) GetExistingRoleBindingsForRole(roleNamespace, role string) ([]*authorizationapi.RoleBinding, error) {
+func (a LocalRoleBindingAccessor) GetExistingRoleBindingsForRole(roleNamespace, role, rb string) ([]*authorizationapi.RoleBinding, error) {
```
Callers upstream should just filter by names themselves

mrogers950 commented:
Changed to filtering by name in AddRole(), leaving GetExistingRoleBindingsForRole() as-is.

liggitt commented on pkg/cmd/admin/policy/modify_roles.go:
```diff
@@ -371,9 +390,16 @@ subjectCheck:
 
 	if isUpdate {
+		if len(o.RoleBindingName) > 0 {
+			roleBinding.Name = o.RoleBindingName
```
setting the name here is wrong... it's either a no-op or it will fail trying to mutate a name

liggitt commented on pkg/cmd/admin/policy/modify_roles.go:
```diff
@@ -336,6 +342,19 @@ func (o *RoleModificationOptions) AddRole() error {
 	if err != nil {
```
you only need to look up existing rolebindings if they didn't provide a rolebinding name

liggitt commented on pkg/cmd/admin/policy/modify_roles.go:
```diff
@@ -336,6 +342,19 @@ func (o *RoleModificationOptions) AddRole() error {
 	}
+
+	if len(o.RoleBindingName) > 0 {
+		filteredRoleBindings := []*authorizationapi.RoleBinding{}
```
filtering the list to a single item is weird... it also means if the named rolebinding already exists but points to a different roleref, it'll get missed here... that should be an error

liggitt commented on pkg/cmd/admin/policy/modify_roles.go:
```diff
@@ -336,6 +342,19 @@ func (o *RoleModificationOptions) AddRole() error {
+		}
+	}
+
 	roleBindingNames, err := o.RoleBindingAccessor.GetExistingRoleBindingNames()
```
you only need to do this if they didn't provide a name

liggitt commented on pkg/cmd/admin/policy/modify_roles.go:
```diff
@@ -371,9 +390,16 @@ subjectCheck:
 		err = o.RoleBindingAccessor.UpdateRoleBinding(roleBinding)
 	} else {
-		roleBinding.Name = getUniqueName(o.RoleName, roleBindingNames)
+		if len(o.RoleBindingName) > 0 {
```
I'd like to see all this logic hoisted up to the top

```
if name provided {
  look up named binding
  if exists and points to other roleref, error
  if exists, set isUpdate=true
  if doesn't exist, populate object with specified name, set isUpdate=false
} else {
  get bindings to role
  if exists, pick first to modify, set isUpdate=true
  if doesn't exist, get existing names and generate unique name, set isUpdate=false
}
```

everything from that point on can deal with the rolebinding the same way in either case

mrogers950 commented:
@liggitt thanks for the review, I've adjusted the AddRole() logic as you suggested, and added the accessor method GetRoleBinding() in order to look for a rolebinding directly by name.

liggitt commented on pkg/cmd/admin/policy/modify_roles.go:
```diff
@@ -332,27 +338,45 @@ func (o *RoleModificationOptions) Complete(f *clientcmd.Factory, args []string,
+	isUpdate := false
+	if len(o.RoleBindingName) > 0 {
+		// Look for an existing rolebinding by name.
+		roleBinding, err = o.RoleBindingAccessor.GetRoleBinding(o.RoleNamespace, o.RoleBindingName)
```
if you get an error other than `NotFound`, return early

liggitt commented on pkg/cmd/admin/policy/modify_roles.go:
```diff
@@ -332,27 +338,45 @@ func (o *RoleModificationOptions) Complete(f *clientcmd.Factory, args []string,
+		// Look for an existing rolebinding by name.
+		roleBinding, err = o.RoleBindingAccessor.GetRoleBinding(o.RoleNamespace, o.RoleBindingName)
+		if roleBinding != nil {
+			if roleBinding.RoleRef.Name != o.RoleName {
```
or if the roleRef namespace is different

liggitt commented on pkg/cmd/admin/policy/policy.go:
```diff
@@ -165,6 +166,21 @@ func (a LocalRoleBindingAccessor) GetExistingRoleBindingNames() (*sets.String, e
 	return ret, nil
 }
 
+func (a LocalRoleBindingAccessor) GetRoleBinding(roleNamespace, roleBindingName string) (*authorizationapi.RoleBinding, error) {
```
`namespace, name string`

liggitt commented on pkg/cmd/admin/policy/policy.go:
```diff
@@ -165,6 +166,21 @@ func (a LocalRoleBindingAccessor) GetExistingRoleBindingNames() (*sets.String, e
 }
 
+func (a LocalRoleBindingAccessor) GetRoleBinding(roleNamespace, roleBindingName string) (*authorizationapi.RoleBinding, error) {
+	roleBindings, err := a.Client.PolicyBindings(a.BindingNamespace).Get(authorizationapi.GetPolicyBindingName(roleNamespace), metav1.GetOptions{})
```
why is this not just `a.Client.RoleBindings(namespace).Get(name, metav1.GetOptions{})`?

liggitt commented on pkg/cmd/admin/policy/policy.go:
```diff
@@ -222,6 +238,22 @@ func (a ClusterRoleBindingAccessor) GetExistingRoleBindingNames() (*sets.String,
 	return ret, nil
 }
 
+func (a ClusterRoleBindingAccessor) GetRoleBinding(roleNamespace, roleBindingName string) (*authorizationapi.RoleBinding, error) {
```
you only need a name

liggitt replied:
I guess you have to conform to the interface, but namespace should always be empty

liggitt commented on pkg/cmd/admin/policy/policy.go:
```diff
@@ -222,6 +238,22 @@ func (a ClusterRoleBindingAccessor) GetExistingRoleBindingNames() (*sets.String,
 }
 
+func (a ClusterRoleBindingAccessor) GetRoleBinding(roleNamespace, roleBindingName string) (*authorizationapi.RoleBinding, error) {
+	uncast, err := a.Client.ClusterPolicyBindings().Get(authorizationapi.GetPolicyBindingName(roleNamespace), metav1.GetOptions{})
```
`return a.Client.ClusterRoleBindings().Get(name, metav1.GetOptions{})`?

liggitt replied:
I guess you have to convert the result in non-error cases

mrogers950 commented:
I've pushed an update to use a.Client.[Cluster]RoleBindings().Get() for the GetRoleBinding implementations instead of searching through all of the policy bindings. As a result the tests needed changes to the fake client reactor functions, and I added a cluster rolebinding test to confirm the GetRoleBinding method for cluster roles.
The revised logic for AddRole should also now properly handle updates and creates both with and without the optional rolebinding name (I realized that after the previous update, the tests passed but manual testing failed).

enj commented on pkg/cmd/admin/policy/modify_roles.go:
```diff
@@ -332,23 +338,43 @@ func (o *RoleModificationOptions) Complete(f *clientcmd.Factory, args []string,
-		isUpdate = false
+	var err error
+	isUpdate := false
+	if len(o.RoleBindingName) > 0 {
```
Add comments for each of the branches in this flow.

enj commented on pkg/cmd/admin/policy/modify_roles_test.go:
```diff
@@ -0,0 +1,319 @@
+			t.Errorf("%s: unexpected err %v", tcName, err)
+		}
+
+		if len(tc.rolebindingname) < 1 {
```
wat?

enj commented on pkg/cmd/admin/policy/modify_roles_test.go:
```diff
@@ -0,0 +1,319 @@
+	kapi "k8s.io/kubernetes/pkg/api"
+)
+
+func TestModifyRoleBindingName(t *testing.T) {
```
In general I have a hard time following this test...

mrogers950 commented:
@enj Rebased, removing the Policy objects in the tests and updated the fakeclient functions accordingly. I also made some renames and added comments to the tests so hopefully it is easier to follow.

mrogers950 commented:
/retest

enj commented:
/test end_to_end

mrogers950 commented:
After Mo's in-person feedback, I've refactored the revised AddRole logic and simplified the tests. Also rebased.

mrogers950 commented:
/test extended_conformance_install_update

mrogers950 commented:
Flake https://github.com/openshift/origin/issues/11114

mrogers950 commented:
/retest

enj commented:
/lgtm

enj commented:
@fabianofranz can you approve this?

mrogers950 commented:
flake #13966

mrogers950 commented:
/retest

juanvallejo commented:
/test extended_conformance_install_update

/lgtm

enj commented:
Approving based on feedback from @juanvallejo

mrogers950 commented:
/retest
